### PR TITLE
[hierarchical-compilation][inductor] Support invoke_subgraph HOP

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #137808
* __->__ #138031
* #137965
* #138036
* #137538

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

File: torch/_inductor/codegen/wrapper.py [hunk 1]
```diff
@@ -2086,6 +2086,14 @@ def codegen_subgraph(self, subgraph, outer_inputs, outer_outputs):
 
         self.codegen_subgraph_suffix(subgraph, outer_inputs, outer_outputs)
 
+    def codegen_invoke_subgraph(self, invoke_subgraph):
```
Comment (aakhundov) [inline torch/_inductor/codegen/wrapper.py:2089]: Should we add this to the `cpp_wrapper_cpu` with `NotImplementedException`?

Comment (anijain2305) [reply torch/_inductor/codegen/wrapper.py:2089]: Good catch.

File: torch/_inductor/ir.py [hunk 2]
```diff
@@ -6528,6 +6528,71 @@ def _has_aliased_buffers(buffers: Sequence[IRNode]) -> bool:
     return len(OrderedSet(id(buffer) for buffer in buffers)) < len(buffers)
 
 
+@ir_dataclass
+class InvokeSubgraph(ExternKernel):
```
Comment (eellison) [inline torch/_inductor/ir.py:6532]: Do we have anything in place that will require similar aliasing relationships ? would you mind adding a test that shows we dont try to say inplace an input in the subgraph ? 

Or similarly, invoke a subgraph once with the same buffer appearing twice, and then with the subgraph invoked and two different buffers, and check that we dont dedupe the tensor arg internally.

File: torch/_inductor/ir.py [hunk 3]
```diff
@@ -6528,6 +6528,71 @@ def _has_aliased_buffers(buffers: Sequence[IRNode]) -> bool:
...
+        self.subgraph = subgraph
+        self.name = V.graph.register_buffer(self)
+        V.graph.register_operation(self)
+
+    @classmethod
+    def create(cls, subgraph: Subgraph, operands):
+        operands = [cls.realize_input(x) for x in operands]
+
+        fx_operands = V.graph.current_node.args[-1]
+        fake_operands = [x.meta["val"] for x in fx_operands]  # type: ignore[union-attr]
```
Comment (eellison) [inline torch/_inductor/ir.py:6554]: inductor is not required to match strides for all intermediaries. We could try to force strides to be equal for these inputs, or we can update the fake_operands to match the strides of the operands input. Generally, it's better to let inductor have flexibility. I guess this also an existing issue with Cond and While loop cc @aakhundov .

If the strides are potentially changing, we may not always be able to reuse the subgraph lowering and would need to do some checking.

**Review by eellison — COMMENTED** [review]
Looks good. Couple questions on subgraph reuse when invoking with different sets of tensors. I think we should fix the stride issue I mentioned. 

Also, general question, where are the sanity checks on when a subgraph can be reused for different inputs ? Say, different shapes, etc. is that higher in the stack ?

File: torch/_inductor/ir.py (hunk 2 above)
Comment (anijain2305) [reply torch/_inductor/ir.py:6532]: https://github.com/pytorch/pytorch/pull/138036 is for aliasing and input mutation.

Dynamo will ensure to not even create an `invoke_subgraph` HOP in fx graph if there is mutation.

File: torch/_inductor/ir.py (hunk 3 above)
Comment (eellison) [reply torch/_inductor/ir.py:6554]: This is also a reason to implement assert_sizes_strides haha - would manifest as silent incorrectness rn if we had divergence

Comment (anijain2305) [reply torch/_inductor/ir.py:6554]: Yes, Horace also mentioned. I am planning to add asserts back for the subgraph

Comment (anijain2305) [reply torch/_inductor/ir.py:6617]: Added code to match the stride. Please check.

File: torch/_inductor/ir.py [hunk 4]
```diff
@@ -6593,6 +6593,80 @@ def _has_aliased_buffers(buffers: Sequence[IRNode]) -> bool:
...
+        # Realize the inputs. Also intermediates can have different strides than
+        # the inputs of the subgraph. So, force the intermediates to have same
+        # strides as that of subgraph inputs.
+        operands = [cls.realize_input(x) for x in operands]
+        # TODO(anijain2305) - Support symint inputs in future.
+        fake_strides = [fake_operand.stride() for fake_operand in fake_operands]
+        operands = [
+            cls.require_exact_strides(x, fake_strides[idx])
+            for idx, x in enumerate(operands)
+        ]
```
Comment (anijain2305) [inline torch/_inductor/ir.py:6628]: @eellison This is the change you requested. I am not sure if this is the right way.

File: torch/_inductor/ir.py [hunk 5]
```diff
@@ -6593,6 +6593,80 @@ def _has_aliased_buffers(buffers: Sequence[IRNode]) -> bool:
...
+    def create(cls, subgraph: Subgraph, operands):
+        fx_operands = V.graph.current_node.args[-1]
+        fake_operands = [x.meta["val"] for x in fx_operands]  # type: ignore[union-attr]
+
+        # Realize the inputs. Also intermediates can have different strides than
+        # the inputs of the subgraph. So, force the intermediates to have same
+        # strides as that of subgraph inputs.
+        operands = [cls.realize_input(x) for x in operands]
+        # TODO(anijain2305) - Support symint inputs in future.
+        fake_strides = [fake_operand.stride() for fake_operand in fake_operands]
```
Comment (eellison) [inline torch/_inductor/ir.py:6624]: we expect sym expr inside inductor, not symints.. it's a bit confusing, something we should fix. 

Can you convert the symints first:

https://github.com/pytorch/pytorch/blob/c272526ea54b8063c596addb48cd3a87ab91a5ef/torch/_inductor/graph.py#L1490-L1493

Also -`require_exact_strides` doesn't work great with aliases/NonOwningLayout.... this should mostly work for now though.

**Review by eellison — APPROVED** [review]

**Pull request closed without merging**